pm/sender_test.go

```diff
@@ -0,0 +1,249 @@
+	sender := defaultSender(t)
+	recipient := randAddressOrFatal(t)
+	ticketParams := defaultTicketParams(t)
+	sessionID := sender.StartSession(recipient, ticketParams)
+	am := sender.accountManager.(*stubAccountManager)
+	am.signShouldFail = true
+
+	_, _, _, err := sender.CreateTicket(sessionID)
+
+	if err == nil {
+		t.Errorf("expected an error when trying to sign the ticket")
+	}
+	if !strings.Contains(err.Error(), "error signing") {
+		t.Errorf("expected error to contain 'error signing' but instead got: %v", err.Error())
+	}
+}
+
+func TestCreateTicket_GivenConcurrentCallsForSameSession_SenderNonceIncrementsCorrectly(t *testing.T) {
+	totalTickets := 100
+	lock := sync.RWMutex{}
+	sender := defaultSender(t)
+	recipient := randAddressOrFatal(t)
+	ticketParams := defaultTicketParams(t)
+	sessionID := sender.StartSession(recipient, ticketParams)
+
+	var wg sync.WaitGroup
+	wg.Add(totalTickets)
+	var tickets []*Ticket
+	for i := 0; i < totalTickets; i++ {
+
+		go func() {
+			ticket, _, _, _ := sender.CreateTicket(sessionID)
+
+			lock.Lock()
+			tickets = append(tickets, ticket)
+			lock.Unlock()
+
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	sessionUntyped, ok := sender.sessions.Load(sessionID)
+	if !ok {
+		t.Fatalf("failed to find session with ID %v", sessionID)
+	}
+	session := sessionUntyped.(*session)
+	if session.senderNonce != uint64(totalTickets) {
+		t.Errorf("expected end state SenderNonce %d to be %d", session.senderNonce, totalTickets)
+	}
+
+	uniqueNonces := make(map[uint64]bool)
+	for _, ticket := range tickets {
+		uniqueNonces[ticket.SenderNonce] = true
+	}
+	if len(uniqueNonces) != totalTickets {
+		t.Errorf("expected unique nonces count %d to be %d", len(uniqueNonces), totalTickets)
+	}
+}
+
+func defaultSender(t *testing.T) *sender {
+	account := accounts.Account{
+		Address: randAddressOrFatal(t),
+	}
+	am := &stubAccountManager{
+		account: account,
+	}
+	s := NewSender(am)
+	return s.(*sender)
+}
+
+func defaultTicketParams(t *testing.T) TicketParams {
+	recipientRandHash := randHashOrFatal(t)
+	return TicketParams{
+		FaceValue:         big.NewInt(0),
+		WinProb:           big.NewInt(0),
+		Seed:              big.NewInt(0),
+		RecipientRandHash: recipientRandHash,
+	}
+}
+
+func randHashOrFatal(t *testing.T) ethcommon.Hash {
+	key, err := randBytes(32)
+
+	if err != nil {
+		t.Fatalf("failed generating random hash: %v", err)
+		return ethcommon.Hash{}
+	}
+
+	return ethcommon.BytesToHash(key[:])
+}
+
+func randAddressOrFatal(t *testing.T) ethcommon.Address {
+	key, err := randBytes(addressSize)
+
+	if err != nil {
+		t.Fatalf("failed generating random address: %v", err)
+		return ethcommon.Address{}
+	}
+
+	return ethcommon.BytesToAddress(key[:])
+}
+
+func randBytesOrFatal(size int, t *testing.T) []byte {
+	res, err := randBytes(size)
+
+	if err != nil {
+		t.Fatalf("failed generating random bytes: %v", err)
+		return nil
+	}
+
+	return res
+}
+
+func randBytes(size int) ([]byte, error) {
```

Comment:
Curious why this implementation was taken rather than the one here; this is nice because it doesn't return an error, so there's less boilerplate to handle in the above helpers.

https://github.com/livepeer/go-livepeer/blob/94c0238ddb5f73c6142d5a622d1de1a6fe1dd730/core/streamdata.go#L49-L55

Reply:
No reason except my not being aware of this implementation. Practically speaking if I really wanted to reuse this code I would have to refactor it to a super basic package that all packages can import, because in the current state if I just used it in `pm` it would result in a dependency cycle.

What do you think about creating such a package, and how would you like us to capture all these little refactoring ideas that come up here?

Reply:
Ah I was referring more to the construct of generating the bytes via iterated `rand.Uint32` rather than `rand.Read` since the former doesn't throw an error. That seems like it would simplify a lot of the invocations of `randBytes`. Agree that trying to use `core.RandomIDGenerator` as-is would be complex. We actually have three or four random-bytes helpers scattered through the code 🙃 

I like the idea of a package for helpers, but it seems like `common` should mostly already be "that" package. The DB stuff is an interesting exception there, but I have some further thoughts which I'll elaborate in the [other thread](https://github.com/livepeer/go-livepeer/pull/655#discussion_r244243772).

Reply:
I'm still of the opinion that we would benefit from separating DB code out of `common`, and would still love to keep the `pm` package as decoupled as possible from any Livepeer-specific assumptions, so this implementation could potentially be reused as an open source package later.

To summarize I am taking the conversation here and in the other thread as potential refactoring TODOs between MVS and the next milestone, since I'd love to discuss it a bit further before we land on a specific action plan.